datadog_lambda/cold_start.py

```diff
@@ -21,3 +25,196 @@ def is_cold_start():
 def get_cold_start_tag():
     """Returns the cold start tag to be used in metrics"""
     return "cold_start:{}".format(str(is_cold_start()).lower())
+
+
+class ImportNode(object):
+    def __init__(self, module_name, full_file_path, start_time_ns, end_time_ns=None):
+        self.module_name = module_name
+        self.full_file_path = full_file_path
+        self.start_time_ns = start_time_ns
+        self.end_time_ns = end_time_ns
+        self.children = []
+
+
+root_nodes: List[ImportNode] = []
+import_stack: List[ImportNode] = []
+already_wrapped_loaders = set()
+
+
+def reset_node_stacks():
+    global root_nodes
+    root_nodes = []
+    global import_stack
+    import_stack = []
+
+
+def push_node(module_name, file_path):
+    node = ImportNode(module_name, file_path, time.time_ns())
+    global import_stack
+    if import_stack:
+        import_stack[-1].children.append(node)
+    import_stack.append(node)
+
+
+def pop_node(module_name):
+    global import_stack
+    if not import_stack:
+        return
+    node = import_stack.pop()
+    if node.module_name != module_name:
+        return
+    end_time_ns = time.time_ns()
+    node.end_time_ns = end_time_ns
+    if not import_stack:  # import_stack empty, a root node has been found
+        global root_nodes
+        root_nodes.append(node)
+
+
+def wrap_exec_module(original_exec_module):
+    def wrapped_method(module):
+        should_pop = False
+        try:
+            spec = module.__spec__
+            push_node(spec.name, spec.origin)
+            should_pop = True
+        except Exception:
+            pass
+        try:
+            return original_exec_module(module)
+        finally:
+            if should_pop:
+                pop_node(spec.name)
+
+    return wrapped_method
+
+
+def wrap_find_spec(original_find_spec):
+    def wrapped_find_spec(*args, **kwargs):
+        spec = original_find_spec(*args, **kwargs)
+        if spec is None:
+            return None
+        loader = getattr(spec, "loader", None)
+        if (
+            loader is not None
+            and isinstance(loader, Hashable)
+            and loader not in already_wrapped_loaders
+        ):
+            if hasattr(loader, "exec_module"):
+                try:
+                    loader.exec_module = wrap_exec_module(loader.exec_module)
+                    already_wrapped_loaders.add(loader)
+                except Exception:
+                    pass
```

Comment:
Do you think we want to debug log here?

Reply:
Yes! I was thinking the same. I was avoiding unnecessary loggings but I think this one is actually useful. For example, if customer did have some custom importer and failed to wrap.